examples/vision/metric_learning.py

```diff
@@ -22,290 +24,380 @@
+                 normalize=False)
 
 """
-As a visual check of these embeddings we can build a collage of the near neighbours for 5
-random examples. The first column of the image below is a randomly selected image, the
-following 10 columns show the nearest neighbours in order of similarity.
-"""
+## No Match
 
-num_collage_examples = 5
-
-examples = np.empty(
-    (
-        num_collage_examples,
-        near_neighbours_per_example + 1,
-        height_width,
-        height_width,
-        3,
-    ),
-    dtype=np.float32,
-)
-for row_idx in range(num_collage_examples):
-    examples[row_idx, 0] = x_test[row_idx]
-    anchor_near_neighbours = reversed(near_neighbours[row_idx][:-1])
-    for col_idx, nn_idx in enumerate(anchor_near_neighbours):
-        examples[row_idx, col_idx + 1] = x_test[nn_idx]
-
-show_collage(examples)
+We can plot the examples outside of the calibrated threshold to see which images
+are not matching any indexed examples.
 
+This may provide insight into what other examples may need to be indexed or
+surface anomalous examples within the class.
 """
-We can also get a quantified view of the performance by considering the correctness of
-near neighbours in terms of a confusion matrix.
 
-Let us sample 10 examples from each of the 10 classes and consider their near neighbours
-as a form of prediction; that is, does the example and its near neighbours share the same
-class?
+idx_no_match = np.where(np.array(matches) == 10)
+no_match_queries = x_confusion[idx_no_match]
+if len(no_match_queries):
+    plt.imshow(no_match_queries[0])
+else:
+    print("All queries have a match below the distance threshold.")
 
-We observe that each animal class does generally well, and is confused the most with the
-other animal classes. The vehicle classes follow the same pattern.
 """
+# Visualize clusters
 
-confusion_matrix = np.zeros((num_classes, num_classes))
+One of the best ways to quickly get a sense of the quality of how the model is
+doing and understand it's short comings is to project the embedding into a 2D
+space.
 
-# For each class.
-for class_idx in range(num_classes):
-    # Consider 10 examples.
-    example_idxs = class_idx_to_test_idxs[class_idx][:10]
-    for y_test_idx in example_idxs:
-        # And count the classes of its near neighbours.
-        for nn_idx in near_neighbours[y_test_idx][:-1]:
-            nn_class_idx = y_test[nn_idx]
-            confusion_matrix[class_idx, nn_class_idx] += 1
+This allows us to inspect clusters of images and understand which classes are
+entangled.
+"""
 
-# Display a confusion matrix.
-labels = [
-    "Airplane",
-    "Automobile",
-    "Bird",
-    "Cat",
-    "Deer",
-    "Dog",
-    "Frog",
-    "Horse",
-    "Ship",
-    "Truck",
-]
-disp = ConfusionMatrixDisplay(confusion_matrix=confusion_matrix, display_labels=labels)
-disp.plot(include_values=True, cmap="viridis", ax=None, xticks_rotation="vertical")
-plt.show()
+num_examples_to_clusters = 720
```

Comment:
720 is a weird number there are 10 classes so 700?

Reply:
Done